### ISSUE TITLE
Performance regression

### ISSUE THREAD

**Issue by mindplay-dk** [yisar/fre#92]
With changes in #86, I noticed a drastic performance regression. 🤔

![image](https://user-images.githubusercontent.com/103348/67617259-a3d72900-f7e1-11e9-9d18-68f9fb771622.png)

![image](https://user-images.githubusercontent.com/103348/67617271-bcdfda00-f7e1-11e9-8574-4d18740c95c9.png)

Could it be that it was just benchmarking substantially better before because it wasn't deallocating memory - and thereby wasn't benchmarking all the memory shuffling it actually has to do?

It doesn't seem like that can reasonably account for a 2-3 x slowdown though? (in some cases more.)

**Comment by 132yse** [comment]
I am not sure what's causing the slowdown, probably because I've rewritten commitwork, which theoretically slows down, but uses less memory
However, since fre uses asynchronous rendering, its running time may be much longer than expected, because it only uses idle time but not full time.

**Comment by 132yse** [comment]
According to the benchmark test results, I saw the difference. Clear didn't slow down because clear is still an array in use, the rest of the items slowed down obviously. Maybe I used the while loop.
I need to reassess them now…

**Comment by 132yse** [comment]
hey，I run the tests，but in my computer, It doesn't seem to slow down significantly. Why?
![QQ图片20191026194951](https://user-images.githubusercontent.com/12951461/67619106-fdfbdc80-f829-11e9-9075-7dad6db75b4f.png)

**Comment by mindplay-dk** [comment]
> I am not sure what's causing the slowdown, probably because I've rewritten commitwork, which theoretically slows down, but uses less memory

Yes, as said, there would be some overhead from deallocation and garbage collection - I'd expect that to add a few percent maybe, but not 300%?

> However, since fre uses asynchronous rendering, its running time may be much longer than expected, because it only uses idle time but not full time.

Sure, but you saw the previous result, and we were actually close to react/preact on many of the benchmarks - the scheduling hasn't drastically changed, so I don't think that can really account for such a drastic change either?

> hey，I run the tests，but in my computer, It doesn't seem to slow down significantly. Why?

Huh, that's interesting.

Are you building the benchmark with `build-prod` or `build-dev`? (I used `build-prod`)

**Comment by 132yse** [comment]
Yes, I tried both `build-prod` and `build-dev`, It's slowing down, but only about 100ms.

The new commitwork enables the update phase to start with subcomponent, Instead of starting with rootFiber( React did ).

Although it slows down the first rendering time, we rarely have more than 1000 elements per component, right?

But I admit that we should do more work on performance optimization in the future, including two core algorithms.

**Comment by 132yse** [comment]
Maybe we can change different computers and browsers for testing, or can we tell me your browser version, and I'm worried about missing any bugs.

**Comment by 132yse** [comment]
I think about it. I think I can improve this algorithm, use breadth first traversal instead of depth first traversal, wait for me.😏

**Comment by mindplay-dk** [comment]
I'm always on the latest version of Chrome.

And yes, I also thought this result looked suspicious, so I immediately re-tested with the `preact` benchmark, to see if results in my browser had changed for any reason, but both `react` and `preact` benchmarks performed similar to before.

**Comment by 132yse** [comment]
I see. I'll try to optimize this algorithm first. Its time complexity is indeed higher than that of traversing arrays before, but it can be improved.

**Comment by 132yse** [comment]
I rethink this, and now I've summed up a few things.

### after: while loop for linked list
This is the same as react [here](https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberCommitWork.js#L1080), but react has other performance optimization mechanisms, such as `batch updates` and `updateQueue`, It can reduce updates, so React is a bit faster.

### before: forEach for array
This is the easiest way to do this. It pushes the fibers into a array, and then forEach the array

The time complexity of array traversal is far lower than the loop of linked list loop.

But it also brings a lot of boundary problems, such as memory overflow.

At the same time, it is not conducive to future work.

In conclusion, I need to make a choice now. Is it difficult, or is there any other way?

**Comment by mindplay-dk** [comment]
Good questions. 👀

Maybe hold off on optimizations for now?

The library can be useful without being the fastest - optimization is likely to introduce bugs at this stage (code coverage is still missing for some details in `scheduler` and `reconciler`) so, if it were up to me, I would focus on stability, worry about performance later.

**Comment by 132yse** [comment]
> Maybe hold off on optimizations for now?
Yep, it is now, but in the future we need to optimize pref in different ways, It will take a long time to accumulate.

**Comment by 132yse** [comment]
Before, I was stupid to try to collect arrays in the parent fiber, which caused bugs, but in fact, I only need a [global array](https://github.com/132yse/fre/blob/master/src/reconciler.js#L11
). I apologize for my stupidity. Now it runs at a better performance and is simpler.

**Comment by mindplay-dk** [comment]
It's actually [slightly slower now](https://docs.google.com/spreadsheets/d/1lA3gLTNihlVMKGTBAz-mascxFLVbpg0K-ZEOfFNf3IA/edit?usp=sharing), but I wouldn't worry too much about it at this point - you simplified the code, which makes it easier to understand, which will make it easier to optimize later on, so these are good changes overall. 🙂

**Comment by mindplay-dk** [comment]
(By the way, I [asked](https://twitter.com/mindplaydk/status/1188841086946598914) Jason Miller how he went about optimizing `preact`, and he pointed me to [this tool](https://github.com/thlorenz/deoptigate) which looks amazing.)

**Comment by 132yse** [comment]
I'm sorry. I don't have a clue now. It shouldn't be slower. Right? I have no experience in performance optimization.

**Comment by 132yse** [comment]
I tried to analyze it with chrome performance, as shown in the following figure.
## After
![QQ图片20191029193206](https://user-images.githubusercontent.com/12951461/67763507-d18ecd00-fa82-11e9-85ed-eaf643321672.png)
## Before
![QQ图片20191029193158](https://user-images.githubusercontent.com/12951461/67763528-dd7a8f00-fa82-11e9-91cc-a56db34bd737.png)
I don't know which one is good, but it seems that after is more delicate.

**Comment by yisar** [comment]
After  simplify and refactor, it will be better now, but I can't figure out the black hole of timing arrangement. There is still a lot of work to do.

**Comment by yisar** [comment]
I have new ideas about this issue。

1. I implemented an new diff algorithm that traverses both ends, which can handle common prefixes and suffixes just like ivi / inferno / snabbdom …

2. I will refactor the scheduling and use the timeout queue to schedule updates instead of priority and expiration time.

Are you still interested? I think fre2 and react are very different, but I think it is worth the effort.

**Comment by yisar** [comment]
![QQ图片20210102172335](https://user-images.githubusercontent.com/12951461/103454455-53315900-4d1f-11eb-8857-fd9cc9090a00.jpg)

Now the performance of fre2 has been greatly improved, which is equal to vue. Although there is still a lot of room for optimization, who knows? Let's continue to explore.

**Comment by yisar** [comment]
![v2-0b63d138b3bf5be29f9d2678d03bb826_720w](https://user-images.githubusercontent.com/12951461/120959812-2a5d3600-c78d-11eb-9953-a2f4cc958689.jpg)

I refactor the core algorithm again, and now it uses `Offscreen rendering`, which is just as exciting as Concurrent rendering.

The core Offscreen rendering rendering is centralized paints, which operates DOM in memory like `DocumentFragment`, and finally paints to the screen at one time.

Now fre is the fastest vdom framework

**Comment by mindplay-dk** [comment]
I know I haven't shown any interest for a while now, but I am still keeping an eye on this project from the sidelines. My job is quite demanding (and fulfilling!) so I don't have much attention left over for side projects.

**Comment by yisar** [comment]
> I know I haven't shown any interest for a while now, but I am still keeping an eye on this project from the sidelines. My job is quite demanding (and fulfilling!) so I don't have much attention left over for side projects.

That’s great. Congratulations on doing a job you like. I also graduated from school and I’m also working now. It’s just to iterate over fre during vacation time
